File: twidere/src/main/java/org/mariotaku/twidere/loader/support/ParcelableStatusesLoader.java
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.loader.support;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v4.content.AsyncTaskLoader;

import org.mariotaku.twidere.Constants;
import org.mariotaku.twidere.loader.iface.IExtendedLoader;
import org.mariotaku.twidere.model.ParcelableStatus;
import org.mariotaku.twidere.util.collection.NoDuplicatesArrayList;

import java.util.List;

public abstract class ParcelableStatusesLoader extends AsyncTaskLoader<List<ParcelableStatus>>
        implements Constants, IExtendedLoader {

    private final List<ParcelableStatus> mData = new NoDuplicatesArrayList<>();
    private final boolean mFirstLoad;
    private final int mTabPosition;
    private boolean mFromUser;

    public ParcelableStatusesLoader(final Context context, final List<ParcelableStatus> data,
                                    final int tabPosition, final boolean fromUser) {
        super(context);
        mFirstLoad = data == null;
        if (data != null) {
            mData.addAll(data);
        }
        mTabPosition = tabPosition;
        mFromUser = fromUser;
    }

    @Override
    public boolean isFromUser() {
        return mFromUser;
    }

    @Override
    public void setFromUser(boolean fromUser) {
        mFromUser = fromUser;
    }

    protected boolean containsStatus(final long statusId) {
        for (final ParcelableStatus status : mData) {
            if (status.id == statusId) return true;
        }
        return false;
    }

    protected boolean deleteStatus(final List<ParcelableStatus> statuses, final long status_id) {
        if (statuses == null || statuses.isEmpty()) return false;
        boolean result = false;
        for (int i = statuses.size() - 1; i >= 0; i--) {
            if (statuses.get(i).id == status_id) {
                statuses.remove(i);
                result = true;
            }
        }
        return result;
    }

    @Nullable
    protected List<ParcelableStatus> getData() {
        return mData;
    }

    protected int getTabPosition() {
        return mTabPosition;
    }

    protected boolean isFirstLoad() {
        return mFirstLoad;
    }

    @Override
    protected void onStartLoading() {
        forceLoad();
    }


}

File: twidere/src/main/java/org/mariotaku/twidere/loader/ParcelableStatusesLoader.kt
/*
 * 				Twidere - Twitter client for Android
 * 
 *  Copyright (C) 2012-2014 Mariotaku Lee <mariotaku.lee@gmail.com>
 * 
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 * 
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 * 
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.mariotaku.twidere.loader

import android.content.Context
import android.support.v4.content.AsyncTaskLoader
import android.text.TextUtils

import org.mariotaku.twidere.Constants
import org.mariotaku.twidere.loader.iface.IExtendedLoader
import org.mariotaku.twidere.model.ParcelableStatus
import org.mariotaku.twidere.util.collection.NoDuplicatesArrayList

abstract class ParcelableStatusesLoader(
        context: Context,
        adapterData: List<ParcelableStatus>?,
        protected val tabPosition: Int,
        private var fromUser: Boolean) : AsyncTaskLoader<List<ParcelableStatus>>(context), Constants, IExtendedLoader {

    protected val data = NoDuplicatesArrayList<ParcelableStatus>()
    protected val isFirstLoad: Boolean

    init {
        isFirstLoad = adapterData == null
        if (adapterData != null) {
            data.addAll(adapterData)
        }
    }

    override fun isFromUser(): Boolean {
        return fromUser
    }

    override fun setFromUser(fromUser: Boolean) {
        this.fromUser = fromUser
    }

    protected fun containsStatus(statusId: String): Boolean {
        for (status in this.data) {
            if (TextUtils.equals(status.id, statusId)) return true
        }
        return false
    }

    protected fun deleteStatus(statuses: MutableList<ParcelableStatus>?, statusId: String): Boolean {
        if (statuses == null || statuses.isEmpty()) return false
        var result = false
        for (i in statuses.indices.reversed()) {
            if (TextUtils.equals(statuses[i].id, statusId)) {
                statuses.removeAt(i)
                result = true
            }
        }
        return result
    }

    override fun onStartLoading() {
        forceLoad()
    }


}
